extends Node2D
## ğŸˆ å•ä¸ªæ°”çƒé€»è¾‘
## åŠŸèƒ½ï¼š
## - é€šè¿‡ blow() è¢«å¤–éƒ¨æ‰“æ°”
## - ç¼©æ”¾åˆ°æœ€å¤§æ—¶ï¼Œæ°”çƒå‘ä¸Šé£èµ°
## - é£èµ°æ—¶å‘å‡º balloon_finished ä¿¡å·

signal balloon_finished   # æ°”çƒé£èµ°æ—¶é€šçŸ¥å¤–éƒ¨ï¼ˆç®¡ç†å™¨ï¼‰

@onready var sprite_2d: Sprite2D = %Sprite2D          # æ°”çƒå¤–è§‚
@onready var rigid_body_2d: RigidBody2D = %RigidBody2D # æ°”çƒç‰©ç†ä½“

# ç¼©æ”¾å‚æ•°
var start_scale: Vector2 = Vector2(0.25, 0.25)  # åˆå§‹å¤§å°
var max_scale: Vector2 = Vector2(1, 1)          # æœ€å¤§å¤§å°
var inflate_step: Vector2 = Vector2(0.1, 0.1)   # æ¯æ¬¡æ‰“æ°”çš„ç¼©æ”¾æ­¥é•¿

func _ready() -> void:
	# åˆå§‹åŒ–ç¼©æ”¾
	sprite_2d.scale = start_scale
	# åˆå§‹ä¸å—é‡åŠ›å½±å“
	rigid_body_2d.gravity_scale = 0

## ğŸˆ å¤–éƒ¨è°ƒç”¨æ‰“æ°”æ–¹æ³•
func blow() -> void:
	if sprite_2d.scale < max_scale:
		# æ¯æ¬¡æ‰“æ°”å¢åŠ ä¸€å®šç¼©æ”¾
		sprite_2d.scale += inflate_step
		# é˜²æ­¢è¶…è¿‡æœ€å¤§å€¼
		if sprite_2d.scale > max_scale:
			sprite_2d.scale = max_scale

	# å¦‚æœè¾¾åˆ°æœ€å¤§ç¼©æ”¾ï¼Œåˆ™å¯åŠ¨é£èµ°é€»è¾‘
	if sprite_2d.scale == max_scale:
		_launch()

## ğŸˆ æ°”çƒé£èµ°
func _launch() -> void:
	# è®¾ç½®ä¸ºåé‡åŠ›ï¼Œè®©æ°”çƒå¾€ä¸Šé£
	rigid_body_2d.gravity_scale = -1
	# å‘å°„ä¿¡å·ï¼Œå‘Šè¯‰ç®¡ç†å™¨â€œæˆ‘é£èµ°äº†â€
	emit_signal("balloon_finished")
	AudioManager.play_sfx("bi",10.0)
